sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala

```diff
@@ -470,12 +484,27 @@ class TungstenAggregationIterator(
 
       // Process the rest of input rows.
       while (inputIter.hasNext) {
-        val newInput = inputIter.next()
-        numInputRows += 1
-        val groupingKey = groupProjection.apply(newInput)
-        buffer.copyFrom(initialAggregationBuffer)
-        processRow(buffer, newInput)
-        externalSorter.insertKV(groupingKey, buffer)
+        var newHashMap = initHashMap()
```

Comment:
I think you might actually want to re-assign to the original `hashMap` by making it into a `var`. This will let us ensure that `UnsafeAggregationIterator.free()` is able to free all memory after failures.